File: bro-to-app/threaded-simple/receiver.py
#!/usr/bin/python

import time
import broker
import select
import threading

## TAKE NOTE: without the call to select, the subscriber.get() would block the main execution python thread (GIL).
## Using select.select actually defers the waiting to the kernel. that allows our main thread to continue.

def broker_receiver():
    ## demo receiver that is subscribed to the topic "demo/threaded-simple"
    ep = broker.Endpoint()
    subscriber = ep.make_subscriber("demo/threaded-simple")
    ep.listen("127.0.0.1", 9999)
    print("endpoint listening...")

    while True:
        fd_sets = select.select([subscriber.fd()], [], [])
        if not fd_sets[0]:
            print("boom. this is the end.")
        (topic, data) = subscriber.get()
        received_event = broker.bro.Event(data)
        print("received on topic: {}    event name: {}    content: {}".format(topic, received_event.name(), received_event.args()))

## fork away a thread to receive messages via broker
thread = threading.Thread(target=broker_receiver, daemon=True)
thread.start()

while True:
    print("main loop can do stuff \\o/")
    time.sleep(1)

File: bro-to-app/threaded-simple/receiver.cc
#include "broker/broker.hh"
#include "broker/bro.hh"

#include <unistd.h>
#include <future>

using namespace broker;

// demo receiver that is subscribed to the topic "demo/threaded-simple"
void broker_receiver() {
    endpoint ep;
    auto subscriber = ep.make_subscriber( {"demo/threaded-simple"} );
    ep.listen("127.0.0.1", 9999);

    while(true) {
        // block the current thread until a message gets available
        // we could also use select, see the threaded python example
        auto msg = subscriber.get();

        broker::topic topic(std::move(msg.first));
        bro::Event response(std::move(msg.second));
        std::cout << "received on topic: " << topic << "     response event name: " << response.name() << "    content: " << response.args() << std::endl;
    }
}

int main() {

    std::thread(broker_receiver).detach();

    while(true) {
        std::cout << "main loop can do stuff \\o/" << std::endl;
        sleep(3);
    }

    return 0;
}